src/operator/nn/dropout-inl.h

```diff
@@ -187,28 +188,101 @@ class DropoutOp {
                                     const index_t N,
                                     const index_t step,
                                     DType *dropout_out,
-                                    DType *mask_out,
+                                    uint8_t *mask_out,
                                     const DType *input_data,
                                     const real_t pkeep) {
       RNG_KERNEL_LOOP(xpu, DType, id, gen, N, step, {
         const real_t rand_num = static_cast<real_t>(genImpl.uniform());
-        mask_out[i] = mshadow_op::threshold_eq::Map<real_t>(rand_num, pkeep) * (1.0f / pkeep);
-        dropout_out[i] = input_data[i] * mask_out[i];
-      });
+        // mask_out is set per bit position
+        // therefore bitwise shift need to be performed here
+        auto maskIdx = i / 8;
+        auto maskOffset = i % 8;
+        bool maskVal = mshadow_op::threshold_eq::Map<real_t>(rand_num, pkeep);
+        if (maskVal) {
+          // set bit
+          mask_out[maskIdx] |= 1U << maskOffset;
```

Comment:
will this lead to race condition if the same maskIdx is being set by multiple threads? Shall each thread handle at least 8 bits?

Reply:
Good catch. I was thinking of setting the step to 8 but forgot to update it in the macro.

Reply:
After checking into it more, I found ideally this should not happen because `RandGenerator<xpu>::kMinNumRandomPerThread` is 64 and therefore by design the step size inside `LaunchRNG` should be a multiple of 8. But then I looked into that piece of code again and found it looks like a bug in calculating the `step`. Please review my latest change in  src/operator/random/sampler.h and let me know if it makes sense. Thanks.

Reply:
Is this for loop parallelized?

Reply:
In general I do not recommend writing code this way. There is not documenntation nor  guarantee that `kMinNumRandomPerThread` will always be greater than 8 in the future. Nor does the dropout operator document any assumption about the value of `kMinNumRandomPerThread`. The code is delicate and will be broken if some contributor changes `kMinNumRandomPerThread` to values like 4. If there's any assumption, we should add an explicit check so that it won't be broken in the future

Reply:
Fair point. I will refactor this piece of code.